modality-ros-hook/src/hooks.rs

```diff
@@ -151,94 +377,214 @@ impl CapturedTime {
-            let node_name: &'static str = &*(pub_state.node_name.as_str() as *const _);
-            // TODO ref here
-            let direction = MessageDirection::Send { local_publisher_graph_id: pub_state.graph_id };
-            let captured_message = CapturedMessage { kvs, topic_name, node_namespace, node_name, direction };
-            let _called_after_dest = LAST_CAPTURED_MESSAGE.try_with(|lcm| {
-                *lcm.borrow_mut() = Some(captured_message);
-            }).is_err();
-        }
+#[no_mangle]
+unsafe extern "C" fn modality_after_rmw_create_publisher(
+    node_ptr: NodePtr,
+    pub_ptr: *const ros::rmw::publisher,
+    type_support: *const ros::rosidl::message_type_support,
+    topic_name: *const c_char,
+) {
+    let nodes = NODES.load();
+    let Some(node_state) = nodes.get(&node_ptr) else {
+        return;
+    };
+
+    let topic_name_str = String::from_utf8_lossy(CStr::from_ptr(topic_name).to_bytes())
+        .trim()
+        .to_string();
+    if CONFIG.ignored_topics.contains(&topic_name_str) {
+        // shortcut return for ignored topics
+        return;
+    }
+
+    let maybe_schema = RosMessageSchema::from_message_type_support(type_support);
+
+    let mut graph_id = None;
+    let mut c_gid: ros::rmw::gid = std::mem::zeroed(); // good as anything?
+    if ros::rmw::get_gid_for_publisher(pub_ptr, &mut c_gid) == 0 {
+        graph_id = Some(PublisherGraphId(c_gid.data));
+    }
+
+    let mut pub_count = 0;
+    if let Some(message_schema) = maybe_schema {
+        PUBLISHERS.rcu(|pubs| {
+            pub_count = pubs.size() + 1;
+            pubs.insert(
+                pub_ptr as usize,
+                PublisherState {
+                    message_schema: message_schema.clone().flatten(&vec![]),
+                    topic_name: topic_name_str.clone(),
+                    node_namespace: node_state.namespace.clone(),
+                    node_name: node_state.name.clone(),
+                    node_timeline_id: node_state.timeline_id,
+                    graph_id,
+                },
+            )
+        });
+
+        let _ = SEND_CH.send(
+            RmwEvent::create_publisher(
+                node_state.timeline_id,
+                graph_id,
+                topic_name_str,
+                message_schema.namespace,
+                message_schema.name,
+                pub_count,
+            )
+            .into(),
+        );
+    }
+}
+
+#[no_mangle]
+unsafe extern "C" fn modality_after_rmw_destroy_publisher(pub_ptr: PublisherPtr) {
+    if let Some(pub_state) = PUBLISHERS.load().get(&pub_ptr) {
+        let _ = SEND_CH.send(
+            RmwEvent::destroy_publisher(pub_state.node_timeline_id, pub_state.graph_id).into(),
+        );
+    }
+
+    PUBLISHERS.rcu(|pubs| pubs.remove(&pub_ptr));
+}
+
+#[no_mangle]
+unsafe extern "C" fn modality_before_rmw_publish(pub_ptr: PublisherPtr, message: *const c_void) {
```

Comment:
Since this is called before the actual function, we need to do sanity checking on the args.
Probably just NULL checking is sufficient